### PR TITLE
fix KT-7618

### PR DESCRIPTION
**Pull request by MichaelNedzelsky**

### CODE REVIEW

File: ant/src/org/jetbrains/kotlin/ant/KotlinCompilerAdapter.java [hunk 1]
```diff
@@ -87,15 +87,24 @@ public boolean execute() throws BuildException {
...
+            }
+        }
+        else {
+            kotlinc.fillArguments();
         }
 
         javac.log("Running javac...");
 
-        // Javac13#execute passes everything in compileList to javac, which doesn't recognize .kt files
         compileList = filterOutKotlinSources(compileList);
```
Comment (bashor) [inline ant/src/org/jetbrains/kotlin/ant/KotlinCompilerAdapter.java]: why do you filter it twice?

File: ant/src/org/jetbrains/kotlin/ant/KotlinCompilerAdapter.java [hunk 2]
```diff
@@ -87,15 +87,24 @@ public boolean execute() throws BuildException {
...
+
+        if (hasKotlinFilesInSources) {
+            kotlinc.execute();
+            if (!Integer.valueOf(0).equals(kotlinc.getExitCode())) {
+                // Don't run javac if failOnError = false and there were errors on Kotlin sources
+                return false;
+            }
+        }
+        else {
+            kotlinc.fillArguments();
```
Comment (bashor) [inline ant/src/org/jetbrains/kotlin/ant/KotlinCompilerAdapter.java]: what for this?

File: ant/src/org/jetbrains/kotlin/ant/KotlinCompilerAdapter.java (hunk 1 above)
Comment (MichaelNedzelsky) [inline ant/src/org/jetbrains/kotlin/ant/KotlinCompilerAdapter.java]: That was a mistake. Of course it should be `compileList = compileListForJavac;`.
https://github.com/JetBrains/kotlin/commit/e7958c1b10b9efe28f35e0fc1b058b682f150efb

File: ant/src/org/jetbrains/kotlin/ant/KotlinCompilerAdapter.java (hunk 2 above)
Comment (MichaelNedzelsky) [inline ant/src/org/jetbrains/kotlin/ant/KotlinCompilerAdapter.java]: kotlinc.execute() invokes fillArgument(), if execute() is not invoked, then kotlinc.getArgs() will be empty. But ` addRuntimeToJavacClasspath(kotlinc)` uses kotlinc.getArgs().

File: compiler/cli/src/org/jetbrains/kotlin/cli/jvm/K2JVMCompiler.java [hunk 3]
```diff
@@ -200,6 +201,16 @@ else if (arguments.script) {
             }
             else {
                 environment = createCoreEnvironment(rootDisposable, configuration);
+
+                if (messageSeverityCollector.anyReported(CompilerMessageSeverity.ERROR)) {
+                    return INTERNAL_ERROR;
```
Comment (bashor) [inline compiler/cli/src/org/jetbrains/kotlin/cli/jvm/K2JVMCompiler.java]: I don't sure that it should be `INTERNAL_ERROR`

File: compiler/cli/src/org/jetbrains/kotlin/cli/js/K2JSCompiler.java [hunk 4]
```diff
@@ -107,29 +106,38 @@ protected ExitCode doExecute(
...
 
-        if (arguments.outputFile == null) {
-            messageCollector.report(CompilerMessageSeverity.ERROR, "Specify output file via -output", CompilerMessageLocation.NO_LOCATION);
+        if (messageSeverityCollector.anyReported(CompilerMessageSeverity.ERROR)) {
             return ExitCode.INTERNAL_ERROR;
         }
 
+        if (sourcesFiles.isEmpty()) {
+            messageSeverityCollector.report(CompilerMessageSeverity.ERROR, "No source files", CompilerMessageLocation.NO_LOCATION);
+            return COMPILATION_ERROR;
```
Comment (bashor) [inline compiler/cli/src/org/jetbrains/kotlin/cli/js/K2JSCompiler.java]: Maybe we should have another exit code for such cases? e.g. `ILLEGAL_ARGUMENTS_ERROR`.

//cc @udalov

Comment (MichaelNedzelsky) [inline compiler/cli/src/org/jetbrains/kotlin/cli/js/K2JSCompiler.java]: This exit code is used only in `CliCompiler.exe`:
`return severityCollector.anyReported(CompilerMessageSeverity.ERROR) ? COMPILATION_ERROR : code;`
Of course it is possible to add new enum entry into ExitCode but it seems not very useful.

File: ant/src/org/jetbrains/kotlin/ant/KotlinCompilerAdapter.java (hunk 2 above)
Comment (bashor) [inline ant/src/org/jetbrains/kotlin/ant/KotlinCompilerAdapter.java]: cheating

Comment (bashor) [inline ant/src/org/jetbrains/kotlin/ant/KotlinCompilerAdapter.java]: let's write comment about it

Comment (bashor) [inline ant/src/org/jetbrains/kotlin/ant/KotlinCompilerAdapter.java]: Could we cover it in tests, if didn't?

Comment (MichaelNedzelsky) [inline ant/src/org/jetbrains/kotlin/ant/KotlinCompilerAdapter.java]: I'll add necessary tests.

File: compiler/cli/src/org/jetbrains/kotlin/cli/js/K2JSCompiler.java (hunk 4 above)
Comment (bashor) [inline compiler/cli/src/org/jetbrains/kotlin/cli/js/K2JSCompiler.java]: We can fix `CliCompiler.exec` :). Any way I don't like that in the same cases we return different code(`COMPILATION_ERROR ` and `INTERNAL_ERROR`)

Comment (udalov) [inline compiler/cli/src/org/jetbrains/kotlin/cli/js/K2JSCompiler.java]: I don't think another exit code is needed

Comment (MichaelNedzelsky) [inline compiler/cli/src/org/jetbrains/kotlin/cli/js/K2JSCompiler.java]: Ok, I'll replace `COMPILATION_ERROR` with `INTERNAL_ERROR`.

Comment (MichaelNedzelsky) [inline compiler/cli/src/org/jetbrains/kotlin/cli/js/K2JSCompiler.java]: It seems in `K2JVMCompiler.doExecute' `INTERNAL_ERROR` is used only for return from `catch` block. For `K2JSCompiler.doExecute' we have sometimes `INTERNAL_ERROR` and sometimes `COMPILATION_ERROR`.

Comment (bashor) [inline compiler/cli/src/org/jetbrains/kotlin/cli/js/K2JSCompiler.java]: I think we should report `INTERNAL_ERROR` only our (internal) issues(crash, exception etc)

Comment (bashor) [inline compiler/cli/src/org/jetbrains/kotlin/cli/js/K2JSCompiler.java]: Well, let's fix it in separate commit

**Comment by bashor** [comment]
looks good for me

**Comment by bashor** [comment]
Ok

**Comment by MichaelNedzelsky** [comment]
Merged manually.

**Pull request closed without merging**